### PR TITLE
planner: add canonical hasher to take in primitive type directly for hashing.

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/51664

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/cascades/canonical.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
+// Copyright 2024 PingCAP, Inc.
```
Comment (hawkingrei) [inline pkg/planner/cascades/canonical.go:1]: Please run the ```make bazel_prepare```.

**Review by hawkingrei — APPROVED** [review]

File: pkg/planner/cascades/BUILD.bazel [hunk 2]
```diff
@@ -3,6 +3,7 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 go_library(
     name = "cascades",
     srcs = [
+        "canonical.go",
```
Comment (elsa0520) [inline pkg/planner/cascades/BUILD.bazel:6]: The name should be related to hash

**Review by elsa0520 — APPROVED** [review]
LGTM

**Pull request merged**